src/bp/lang-server/service/index.ts

```diff
@@ -45,75 +54,172 @@ export default class LanguageService {
-      this._provideLanguage(langCode, path.join(this.langDir, f))
+      this._provideLanguage(langCode, fastTextModelFileInfo.file, bpeModelFileInfo.file)
     })
   }
 
-  private _provideLanguage(lang: string, modelPath: string) {
+  private _provideLanguage(lang: string, fastTextModelName: string, bpeModelName: string) {
+    const fastTextModelPath = path.join(this.langDir, fastTextModelName)
+    const bpeModelPath = path.join(this.langDir, bpeModelName)
+
     try {
-      if (!fs.existsSync(modelPath)) {
-        throw new Error(`Model '${lang}' at location '${modelPath}' does not exist`)
+      if (!fs.existsSync(fastTextModelPath)) {
+        throw new Error(`Model '${lang}' at location '${fastTextModelPath}' does not exist`)
+      }
+      if (!fs.existsSync(bpeModelPath)) {
+        throw new Error(`Model '${lang}' at location '${bpeModelPath}' does not exist`)
       }
     } catch (err) {
-      throw new VError(`Could not find model '${lang}' at '${modelPath}'`, err)
+      throw new VError(`Could not find model '${lang}' in '${this.langDir}'`, err)
     }
 
-    this._models[lang] = {
+    const fastTextModel = {
       name: lang,
-      path: modelPath,
+      path: fastTextModelPath,
       sizeInMb: 0,
       loaded: false
     }
+
+    const bpeModel = {
+      name: lang,
+      path: bpeModelPath,
+      sizeInMb: 0,
+      loaded: false
+    }
+    this._models[lang] = { fastTextModel, bpeModel }
   }
 
   @WrapErrorsWith(args => `Couldn't load language model "${args[0]}"`)
-  private async _loadModel(lang: string) {
+  private async _loadModels(lang: string) {
+    const fastTextModel = await this._loadFastTextModel(lang)
+    const bpeModel = await this._loadBPEModel(lang)
+    this._models[lang] = { fastTextModel, bpeModel }
+  }
+
+  private async _loadFastTextModel(lang: string): Promise<LoadedFastTextModel> {
     const usedBefore = process.memoryUsage().rss / 1024 / 1024
     const dtBefore = Date.now()
 
     const model = new toolkit.FastText.Model(false, true, true)
-    await model.loadFromFile(this._models[lang].path)
+    await model.loadFromFile(this._models[lang].fastTextModel.path)
 
     const dtAfter = Date.now()
     const usedAfter = process.memoryUsage().rss / 1024 / 1024
     const usedDelta = Math.round(usedAfter - usedBefore)
     const dtDelta = dtAfter - dtBefore
 
-    this._models[lang] = <LoadedModel>{ ...this._models[lang], model, sizeInMb: usedDelta, loaded: true }
+    const loadedModel = <LoadedFastTextModel>{
+      ...this._models[lang].fastTextModel,
+      model,
+      sizeInMb: usedDelta,
+      loaded: true
+    }
     console.log(`language model '${lang}' took about ${usedDelta}mb of RAM and ${dtDelta}ms to load`)
+
+    return loadedModel
   }
 
-  async vectorize(tokens: string[], lang: string): Promise<number[][]> {
-    const _model = this._models[lang] as LoadedModel
-    if (!_model || !_model.loaded) {
+  private async _loadBPEModel(lang: string): Promise<LoadedBPEModel> {
```

Comment:
If you add another method that borrows much code from another similar method, it would make sense to extract that logic into a "generic" method, which is called by the two specific ones. We try to avoid as much as possible to duplicate logic

Reply:
Both methods are actually quite different. I'll make an effort to extract some common logic.